code/modules/mob/living/simple_animal/friendly/corgi.dm

```diff
@@ -487,56 +494,117 @@
+						dir = EAST
+					else if (movement_target.loc.y > src.y)
+						dir = NORTH
+					else
+						dir = SOUTH
+
+					if(isturf(movement_target.loc) && src.Adjacent(movement_target))
+						movement_target.attack_animal(src)
+					else if(ishuman(movement_target.loc))
+						if(prob(20))
+							emote("me", 1, "stares at [movement_target.loc]'s [movement_target] with a sad puppy-face and whimpers.")	
+				ian_status = IDLE
 				movement_target = null
-				stop_automated_movement = 0
-			if( !movement_target || !(movement_target.loc in oview(src, 3)) )
-				movement_target = null
-				stop_automated_movement = 0
-				for(var/obj/item/weapon/reagent_containers/food/snacks/S in oview(src,3))
-					if(isturf(S.loc) || ishuman(S.loc))
-						movement_target = S
-						break
-			if(movement_target)
-				spawn(0)
+							
+		else if(ian_status == BEGIN_POINTER_FOLLOWING)
+			ian_status = POINTER_FOLLOWING
+			if(prob(35) || (master != null && pointer_caller == master))
+				spawn(0) // Separate process
 					stop_automated_movement = 1
-					step_to(src,movement_target,1)
-					sleep(3)
-					step_to(src,movement_target,1)
-					sleep(3)
-					step_to(src,movement_target,1)
-
-					if(movement_target)		//Not redundant due to sleeps, Item can be gone in 6 decisecomds
+					var/failedsteps = 0
+					while(failedsteps <= 3)
+						if(!movement_target || src.Adjacent(movement_target) || get_dist(src, movement_target) >= 7)
+							break
+						if(!step_towards(src,movement_target,1))
+							failedsteps++
+						sleep(6)
+						
+					if(movement_target)		
 						if (movement_target.loc.x < src.x)
 							dir = WEST
 						else if (movement_target.loc.x > src.x)
 							dir = EAST
-						else if (movement_target.loc.y < src.y)
-							dir = SOUTH
 						else if (movement_target.loc.y > src.y)
 							dir = NORTH
 						else
 							dir = SOUTH
-
-						if(isturf(movement_target.loc) && src.Adjacent(movement_target))
-							movement_target.attack_animal(src)
-						else if(ishuman(movement_target.loc) )
-							if(prob(20))
-								emote("me", 1, "stares at [movement_target.loc]'s [movement_target] with a sad puppy-face")
-//PC stuff-Sieve
-
+						step_towards(src,movement_target,1)
+						playsound(loc, 'sound/voice/corgibark.ogg', 80, 1)
+						if(istype(movement_target,/obj/item/weapon/reagent_containers/food/snacks))
+							emote("me", 1, "barks at [movement_target], as if begging it to go into his mouth.")	
+							ian_status = BEGIN_FOOD_HUNTING
+						else if(ishuman(movement_target))
+							emote("me", 1, "barks at [movement_target] and wags his tail.")	
+							ian_status = IDLE
+						else
+							emote("me", 1, "barks with an attitude!")
+							ian_status = IDLE
+								
+			else
+				emote("me", 1, "stares into space with a blank expression.")
+				ian_status = IDLE
+			
+/mob/living/simple_animal/corgi/Ian/proc/get_target()
+	var/vision_range = 5
+	for(var/obj/item/weapon/reagent_containers/food/snacks/S in oview(src,vision_range))
```

Comment:
Optimization: Rather than calling view/oview 3 times, have a var/list/can_see = view(src, vision_range)) and just look through that

Reply:
Yeah, I'll do that optimization. I'm a little busy with some real-life stuff right now and I found a bug where food with a plate bugs his AI out, so it'll be a bit.